flink-table/flink-table-common/src/main/java/org/apache/flink/table/descriptors/TableDescriptor.java

```diff
@@ -19,11 +19,132 @@
 package org.apache.flink.table.descriptors;
 
 import org.apache.flink.annotation.PublicEvolving;
+import org.apache.flink.table.api.ValidationException;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nullable;
+
+import java.util.Map;
+
+import static org.apache.flink.table.descriptors.StreamTableDescriptorValidator.UPDATE_MODE;
+import static org.apache.flink.table.descriptors.StreamTableDescriptorValidator.UPDATE_MODE_VALUE_APPEND;
+import static org.apache.flink.table.descriptors.StreamTableDescriptorValidator.UPDATE_MODE_VALUE_RETRACT;
+import static org.apache.flink.table.descriptors.StreamTableDescriptorValidator.UPDATE_MODE_VALUE_UPSERT;
 
 /**
-  * Common class for all descriptors describing table sources and sinks.
-  */
+ * Describes a table consisting of a connector (in a given update mode) and a format.
+ */
 @PublicEvolving
-public abstract class TableDescriptor extends DescriptorBase {
+public abstract class TableDescriptor<D extends TableDescriptor<D>> extends DescriptorBase {
```

Comment:
Do we need both `TableDescriptor` & `ConnectTableDescriptor`?

Reply:
TableDescriptor is used in `CatalogTableBuilder`. The main reason of this class is the generic that is not present in `ConnectTableDescriptor`.